### DISCUSSION TITLE
tutorials/lift-off-part4/09-our-mutation-in-the-browser

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#45] [course feedback]
# tutorials/lift-off-part4/09-our-mutation-in-the-browser

https://www.apollographql.com/tutorials/lift-off-part4/09-our-mutation-in-the-browser

**Comment by krutilins** [comment]
Hi, how does apollo client define which properties should be compared to define if exactly object we need was changed? 

In our case, if I understand properly

1. what we are doing is requesting `track` and storing in the cache
2. performing mutation and then based on the returned `id` keyword define what should be changed

what if I do not request id keyword in query schema?
Is it documented somewhere?

**Reply by mabuyo** [reply]
Hey @krutilins! Great question. By default, Apollo Client uses the object's `__typename` and `id`, separated by a colon `:` to generate the cache ID. Using that cache ID, it keeps track of which object to change in the cache. 

If the cache can't generate a cache ID for a particular object (for example, if no `id` or `_id` field is present), that object is cached directly inside its parent object, and it must be referenced via the parent (this means the cache isn't always completely flat).

For an example about how this works, you can [read more about it in the Apollo documentation](https://www.apollographql.com/docs/react/caching/overview). Hope that helps!

**Comment by zibiv** [comment]
Have I understood correctly, that in our Mutation in fact I can request only id and num of views from the updated Track

    const INCREMENT_TRACK_VIEWS = graphql(`
      mutation Mutation($incrementTrackViewsId: ID!) {
        incrementTrackViews(id: $incrementTrackViewsId) {
          track {
            id
            numberOfViews
          }
        }
      }
    `);
 and result will be the same?  We've got back our updated value for numberOfViews and id of the Track. Apollo Client will update its numberOfViews field in the cache.

**Reply by mabuyo** [reply]
Hey there @zibiv - apologies for the delayed response, our notifications integration stopped working and I completely missed this!

To clarify, in your mutation, you can request any fields from the `Track` object that you might need in the client application! In this case, we want to include the `numberOfViews` because it would have changed from our mutation, and we want to have the page reflect that new value.

Correct, Apollo Client will take care of updating it in the cache. It can do this automatically because we've included the `id` in our mutation response, which will be used as the unique key to identify that particular Track object in the cache.

Hope that helps, happy to clarify anything else although I realize it has been a while.

**Comment by SteveBold** [comment]
Good afternoon from Liverpool. I am winning with all tutorials, but whilst using Chrome and viewing the second track as per the instruction on the first task here, I am not seeing the console output. It is frustrating as I believe I have done everything correct. The console suggested adding React and Apollo extensions, which I have done and I have also ensured I am signed into my google account, however the only output within the console area of the developer tools is:

"Download the Apollo DevTools for a better development experience: https://chrome.google.com/webstore/detail/apollo-client-developer-t/jdkknkkbebbapilgoeccciglkfbmbnfm"

Obviously as I have added the extensions already, I am seeing a question of whether I want to remove the extension.

Any ideas? I will move on, assuming its some weird force of nature against me :-)

**Reply by jkpchang** [reply]
in the previous section, it told you to add `console.log` in `useMutation` hook.

```
  const [incrementTrackViews] = useMutation(INCREMENT_TRACK_VIEWS, {
    variables: { incrementTrackViewsId: id },
    onCompleted: (data) => {
      console.log(data);
    }
  });

```

**Reply by mabuyo** [reply]
So sorry we missed your comment @SteveBold and thank you @jkpchang for chiming in with a solution! Definitely recommend checking that you have the `onCompleted` function with the console log in your code. If that didn't work, can you confirm that the number of views are going up?

**Reply by SteveBold** [reply]
Thank you both, I will check it out once back on line. I can confirm the number of views is increasing. The tutorials are a dream 👍👍

**Comment by BlockBusterk** [comment]
We have learned query and mutation, how about subscription?

**Reply by mabuyo** [reply]
It's on our list of courses to tackle @BlockBusterk !

**Comment by MichaelNovikov95** [comment]
Thats a perfect tutorial for beginners ! 
I will definitely proceed to work with apollo and graphql.
Thanks for your job, guys :)

**Comment by xavier643** [comment]
Originally in my mutation I left out the id inside of of track. I saw little reason to retrieve that information. However, without that being present my number of views where updating but only if I refreshed the page manually. Can you explain to me why the caching was not auto updating unless I pulled the id as well?

**Reply by mabuyo** [reply]
Hey there @xavier643 ! Great question. By default, Apollo Client uses the object's `__typename` and `id`, separated by a colon `:` to generate the cache ID. Using that cache ID, it keeps track of which object to change in the cache. 

If the cache can't generate a cache ID for a particular object (for example, if no `id` or `_id` field is present), that object is cached directly inside its parent object, and it must be referenced via the parent (this means the cache isn't always completely flat). Without the cache ID and without further instructions on _how_ to update the cache, Apollo Client doesn't know that it can update that `numberOfViews` field for that particular track.

For an example about how this works, you can [read more about it in the Apollo documentation](https://www.apollographql.com/docs/react/caching/overview). Hope that helps!

**Comment by dt-aditisharma** [comment]
Hello, I have been working through this course with little to no issues, up until now. I am able to run my client and server without issues, I was also able to run generate command successfully for both server and client, however, I see an error in the browser console: Uncaught ReferenceError: id is not defined. Now I see the red squiggly line on the "id" in track-card.tsx and can't seem to find whats missing or wrong. Can you please help me out with this?

**Reply by dt-aditisharma** [reply]
I was able to fix this in the code.

**Reply by lizhennessy** [reply]
Thanks for the update @dt-aditisharma ! Glad you were able to figure it out. Let us know if you run into any other issues!

**Comment by NickSomsen** [comment]
I don't typically leave comments or reviews, but I just have to say that this Lift-off series is so well put together! I was able to follow along exactly as described without any unexpected errors, and that for all four parts. The instructions are very clear as well as the visuals. I like the interactivity with the  coding challenges and questions while also following along on localhost. The underlining of GraphQL jargon terms with definitions when you hover on them is very handy. The overall look and feel is very nice and I genuinely have no comments on how this series could be improved. Well done!👏🚀🎉

**Reply by mabuyo** [reply]
Thank you so much for this message @NickSomsen - it made my day! 😊 Don't hesitate to leave a comment if you have any feedback for future courses. Excited to see how your GraphQL journey continues!

**Comment by Chip-L** [comment]
This course was great! I wish there was a little more. :) I just read through [deprecated] Full Stack Quickstart tutorial and it went into the caching and managing local state. That was fabulous! I would love to see another module in this serries that would go into that as well. I really struggle with caching and how to manage the local cache.

**Reply by lizhennessy** [reply]
hey there @Chip-L ! we actually have a course on the Apollo Client cache that we should be releasing somewhat soon. I'd be happy to let you know when it's out! In the meantime, we've got many more backend-focused courses, as well as another fullstack series called [Voyage](https://www.apollographql.com/tutorials/voyage-part1) which takes you through GraphQL federation!

**Comment by sailesh97** [comment]
I absolutely loved the course. Whether it's content quality or slides or anything else, it's just perfect.

I've one request though, unrelated to this course:
I was working on ApolloServer migration to v4 recently in my org and struggled to understand the RESTDataSource caching. I was expecting the extensive content mentioned in Readme.md and Release Document of ApolloDataSource in the official documentation of ApolloServer.

Referring to: 
- https://github.com/apollographql/datasource-rest/blob/v6.4.1/README.md
- https://github.com/apollographql/datasource-rest/releases/tag/v5.0.0

The existing caching docs we have for AS4 is very confusing and unclear to me.

I'm still wondering when does cache comes to picture:
- Is it always or only when we get cache headers from the API, our Apollo Server communicating to?
- In this course, we've no where used @cacheControl and still caching worked.
- There's no visual way to validate whether caching working for a query ran in ApolloServer. 
- when to use responseCachePlugin?
- Cache for a GET call vs a POST call
- How it's different from Browser style caching?
- If I'm mentioned cacheControl for few fields in a schema and the response is not getting cached; I don't see any alerts or warnings with the reason why it didn't cache. 

It would be great to dive deep into the nitty gritty details of Apollo RESTDataSource.

Thanks

**Comment by CasperRademakerCode** [comment]
Great course! Have you considered Credly badges for the related certification?

Thanks a lot, best regards,

Casper
